felix/nftables/table.go

```diff
@@ -0,0 +1,1173 @@
+	if chain := t.chainNameToChain[chainName]; chain != nil {
+		chain.Rules = rules
+	}
+
+	// Incref any newly-referenced chains, then decref the old ones.  By incrementing first we
+	// avoid marking a still-referenced chain as dirty.
+	t.maybeIncrefReferredChains(chainName, rules)
+	t.maybeDecrefReferredChains(chainName, oldRules)
+}
+
+// AppendRules sets the rules to be appended to a given non-Calico chain.
+// These rules are always appended, even if chain insert mode is "insert".
+// If chain insert mode is "append", these rules are appended after any
+// rules added with InsertOrAppendRules.
+func (t *nftablesTable) AppendRules(chainName string, rules []generictables.Rule) {
+	t.logCxt.WithField("chainName", chainName).Debug("Updating rule appends")
+	oldRules := t.chainToAppendedRules[chainName]
+	t.chainToAppendedRules[chainName] = rules
+	numRulesDelta := len(rules) - len(oldRules)
+	t.gaugeNumRules.Add(float64(numRulesDelta))
+	t.dirtyBaseChains.Add(chainName)
+
+	// Incref any newly-referenced chains, then decref the old ones.  By incrementing first we
+	// avoid marking a still-referenced chain as dirty.
+	t.maybeIncrefReferredChains(chainName, rules)
+	t.maybeDecrefReferredChains(chainName, oldRules)
+}
+
+func (t *nftablesTable) UpdateChains(chains []*generictables.Chain) {
+	for _, chain := range chains {
+		t.UpdateChain(chain)
+	}
+}
+
+func (t *nftablesTable) UpdateChain(chain *generictables.Chain) {
+	t.logCxt.WithField("chainName", chain.Name).Debug("Adding chain to available set.")
+	oldNumRules := 0
+
+	// Incref any newly-referenced chains, then decref the old ones.  By incrementing first we
+	// avoid marking a still-referenced chain as dirty.
+	t.maybeIncrefReferredChains(chain.Name, chain.Rules)
+	if oldChain := t.chainNameToChain[chain.Name]; oldChain != nil {
+		oldNumRules = len(oldChain.Rules)
+		t.maybeDecrefReferredChains(chain.Name, oldChain.Rules)
+	}
+	t.chainNameToChain[chain.Name] = chain
+	numRulesDelta := len(chain.Rules) - oldNumRules
+	t.gaugeNumRules.Add(float64(numRulesDelta))
+	if t.chainIsReferenced(chain.Name) {
+		t.dirtyChains.Add(chain.Name)
+	}
+}
+
+func (t *nftablesTable) RemoveChains(chains []*generictables.Chain) {
+	for _, chain := range chains {
+		t.RemoveChainByName(chain.Name)
+	}
+}
+
+func (t *nftablesTable) RemoveChainByName(name string) {
+	t.logCxt.WithField("chainName", name).Debug("Removing chain from available set.")
+	if oldChain, known := t.chainNameToChain[name]; known {
+		t.gaugeNumRules.Sub(float64(len(oldChain.Rules)))
+		t.maybeDecrefReferredChains(name, oldChain.Rules)
+		delete(t.chainNameToChain, name)
+		if t.chainIsReferenced(name) {
+			t.dirtyChains.Add(name)
+
+			// Defensive: make sure we re-read the dataplane state before we make updates.  While the
+			// code was originally designed not to need this, we found that other users of
+			// nftables can still clobber out updates so it's safest to re-read the state before
+			// each write.
```

Comment:
comment feels like it was just `s/iptables/nftables/` and may not actually be true